src/bin/pg_autoctl/cli_do_tmux.c

```diff
@@ -439,7 +514,333 @@ cli_do_tmux_script(int argc, char **argv)
+		exit(EXIT_CODE_INTERNAL_ERROR);
+	}
+
+	/* prepare the tmux script */
+	(void) prepare_tmux_script(&options, script, true);
+
+	/* memory allocation could have failed while building string */
+	if (PQExpBufferBroken(script))
+	{
+		log_error("Failed to allocate memory");
+		destroyPQExpBuffer(script);
+
+		exit(EXIT_CODE_INTERNAL_ERROR);
+	}
+
+	fformat(stdout, "%s", script->data);
+
+	destroyPQExpBuffer(script);
+}
+
+
+/*
+ * cli_do_tmux_session starts an interactive tmux session with the given
+ * specifications for a cluster. When the session is detached, the pg_autoctl
+ * processes are stopped.
+ */
+void
+cli_do_tmux_session(int argc, char **argv)
+{
+	TmuxOptions options = tmuxOptions;
+	PQExpBuffer script = createPQExpBuffer();
+
+	char scriptName[MAXPGPATH] = { 0 };
+
+	bool success = true;
+
+	/*
+	 * Write the script to "script-${first-pgport}.tmux" file in the root
+	 * directory.
+	 */
+	log_debug("mkdir -p \"%s\"", options.root);
+	if (!ensure_empty_dir(options.root, 0700))
+	{
+		/* errors have already been logged. */
+		exit(EXIT_CODE_INTERNAL_ERROR);
+	}
+
+	if (!normalize_filename(options.root, options.root, sizeof(options.root)))
+	{
+		/* errors have already been logged. */
+		exit(EXIT_CODE_INTERNAL_ERROR);
+	}
+
+	/*
+	 * Prepare the tmux script.
+	 */
+	if (script == NULL)
+	{
+		log_error("Failed to allocate memory");
+		exit(EXIT_CODE_INTERNAL_ERROR);
+	}
+
+	(void) prepare_tmux_script(&options, script, false);
+
+	/* memory allocation could have failed while building string */
+	if (PQExpBufferBroken(script))
+	{
+		log_error("Failed to allocate memory");
+		destroyPQExpBuffer(script);
+
+		exit(EXIT_CODE_INTERNAL_ERROR);
+	}
+
+	/*
+	 * Write the script to file.
+	 */
+	sformat(scriptName, sizeof(scriptName),
+			"%s/script-%d.tmux", options.root, options.firstPort);
+
+	log_info("Writing tmux session script \"%s\"", scriptName);
+
+	if (!write_file(script->data, script->len, scriptName))
+	{
+		log_fatal("Failed to write tmux script at \"%s\"", scriptName);
+		exit(EXIT_CODE_INTERNAL_ERROR);
+	}
+
+	destroyPQExpBuffer(script);
+
+	/*
+	 * Start a tmux session from the script.
+	 */
+	if (!tmux_start_server(options.root, scriptName))
+	{
+		success = false;
+		log_fatal("Failed to start the tmux session, see above for details");
+	}
+
+	/*
+	 * Stop our pg_autoctl processes and kill the tmux session.
+	 */
+	log_info("tmux session ended: kill pg_autoct processes");
+	success = success && tmux_stop_pg_autoctl(&options);
+	success = success && tmux_kill_session(&options);
+
+	if (!success)
+	{
+		exit(EXIT_CODE_INTERNAL_ERROR);
+	}
+}
+
+
+/*
+ * cli_do_tmux_stop runs pg_autoctl stop on all the pg_autoctl process that
+ * might be running in a tmux session.
+ */
+void
+cli_do_tmux_stop(int argc, char **argv)
+{
+	TmuxOptions options = tmuxOptions;
+
+	/* prepare the XDG environment */
+	if (!tmux_prepare_XDG_environment(options.root))
+	{
+		exit(EXIT_CODE_INTERNAL_ERROR);
+	}
+
+	if (!tmux_stop_pg_autoctl(&options))
```

Comment:
Shouldn't this try to stop the tmux session as well?

Reply:
I have used it as a debugging tool mostly when something went wrong after `C-b d`, I'm not sure we have another use case than “things went wrong, help me”. When things went wrong, being able to `tmux at` and take it from there has been quite helpful to me.

Reply:
There's now `pg_autoctl do tmux clean` that cleans processes, directories and kills the tmux session.